File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
class Solution {
public:
    
    bool c(vector<int> &arr,int mid,int k){
        int n=arr.size();
        long long totalhours=0;
        for (int i=0;i<n;i++){
            totalhours+=(arr[i]+mid-1)/mid;
        }
        return totalhours<=k;
    }
    int minEatingSpeed(vector<int>& piles, int h) {
        int n=piles.size();
        int low=1;
        int high=*max_element(piles.begin(),piles.end());
        long long ans=INT_MIN;
        while(low<=high){
            int mid=low+(high-low)/2;
            if(c(piles,mid,h)){
                ans=mid;
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }
        return ans;
    }
};

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    
    boolean caneat(int[] arr,int k,int h){
        long hours=0;
        for (int i:arr){
            hours+=(i+h-1)/h;
        }
        return hours<=k;
    }
    int m(int[] arr){
        int max=arr[0];
        for(int i=1;i<arr.length;i++){
            if(arr[i]>max){
                max=arr[i];
            }
        }
        return max;
    }
    public int minEatingSpeed(int[] arr, int k) {
        int low=1;
        int high = m(arr);
        while(low<=high){
            int mid=low+(high-low)/2;
            if (caneat(arr,k,mid)){
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }
        return low;
    }
}